### PR TITLE
Replace some desc logic in librustc_lint with article_and_desc

### PR DESCRIPTION
**Pull request by mark-i-m**
r? @eddyb @Centril @matthewjasper 

Followup to https://github.com/rust-lang/rust/pull/69674

Blocked on #69498

### CODE REVIEW

**Comment by mark-i-m** [comment]
I still need to run tests

**Comment by mark-i-m** [comment]
@rustbot modify labels: +S-waiting-on-author -S-blocked

I need to fix the ICE

**Comment by mark-i-m** [comment]
Nevermind, the ICE was all in my head. Tests are passing locally.

@rustbot modify labels: +S-blocked

Blocked on #69498, but otherwise ready

**Comment by mark-i-m** [comment]
It looks like those functions are called for def_ids that have `DefPathData::Misc`, which doesn't really give us enough info. This causes article_and_description to crash.

**Comment by eddyb** [comment]
The important thing for `article_and_description` should be the `DefKind`, not `DefPathData` (which is just a crutch) - can you figure out what those `DefId`s are?
We should start adding `DefKind`s as needed.

**Comment by mark-i-m** [comment]
@rustbot modify labels: +S-waiting-on-author -S-blocked

@eddyb It looks like they are things like `use` items, etc. I will take a closer look soon. A quick fix for this PR might be to check the hir ItemKind and return early for the ones we don't want. Does every item have a `DefId`? i.e. if I have an arbitrary `HirId` can I always get a `DefId` to use with `def_kind`?

**Comment by eddyb** [comment]
> Can you tell me when it would return `None`?

For things that don't have a `DefKind` variant yet (`DefKind` originated in name resolution so it doesn't yet have variants like `Impl` or `Closure`, but we should eventually move everything over).

**Comment by mark-i-m** [comment]
@eddyb This is ready now. ui tests passing locally.

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -349,7 +349,7 @@ impl MissingDoc {
         id: Option<hir::HirId>,
         attrs: &[ast::Attribute],
         sp: Span,
-        desc: &'static str,
+        desc: &str,
```
Comment (eddyb) [inline src/librustc_lint/builtin.rs]: If you pass the `article` and `desc` separately you can avoid needing to allocate a string ahead of time (and you could keep the `'static`).

File: src/librustc_lint/builtin.rs [hunk 2]
```diff
@@ -349,7 +349,7 @@ impl MissingDoc {
         id: Option<hir::HirId>,
```
Comment (eddyb) [inline src/librustc_lint/builtin.rs]: Does anything pass `None` for this? If not you could remove the `Option` and compute `article_and_desc` inside here from the `HirId`.

Comment (mark-i-m) [reply src/librustc_lint/builtin.rs]: Unfortunately, it is passed with a `Crate` down below, so for now I will keep the option.

File: src/librustc_lint/builtin.rs [hunk 3]
```diff
@@ -398,7 +399,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for MissingDoc {
     }
 
     fn check_crate(&mut self, cx: &LateContext<'_, '_>, krate: &hir::Crate<'_>) {
-        self.check_missing_docs_attrs(cx, None, &krate.item.attrs, krate.item.span, "crate");
+        self.check_missing_docs_attrs(cx, None, &krate.item.attrs, krate.item.span, "a", "crate");
```
Comment (eddyb) [inline src/librustc_lint/builtin.rs]: Bonus: you can use definite article (i.e. `"the"`) instead of indefinite.

**Review by eddyb — APPROVED** [review]

**Comment by mark-i-m** [comment]
Oh, it looks like I missed a test and I think bors will fail :(

**Pull request merged**